src/trace/context/extractors/sns-sqs.ts

```diff
@@ -3,32 +3,46 @@ import { EventTraceExtractor } from "../extractor";
 import { TracerWrapper } from "../../tracer-wrapper";
 import { logDebug } from "../../../utils";
 import { SpanContextWrapper } from "../../span-context-wrapper";
+import { XrayService } from "../../xray-service";
 
 export class SNSSQSEventTraceExtractor implements EventTraceExtractor {
   constructor(private tracerWrapper: TracerWrapper) {}
 
   extract(event: SQSEvent): SpanContextWrapper | null {
-    const body = event?.Records?.[0]?.body;
-    if (body === undefined) return null;
-
     try {
-      const parsedBody = JSON.parse(body) as SNSMessage;
-      const messageAttribute = parsedBody?.MessageAttributes?._datadog;
-      if (messageAttribute.Value === undefined) return null;
+      // First try to extract trace context from message attributes
+      if (event?.Records?.[0]?.body) {
+        const parsedBody = JSON.parse(event?.Records?.[0]?.body) as SNSMessage;
+        const messageAttribute = parsedBody?.MessageAttributes?._datadog;
+        if (messageAttribute?.Value) {
+          let headers;
+          if (messageAttribute.Type === "String") {
+            headers = JSON.parse(messageAttribute.Value);
+          } else {
+            const decodedValue = Buffer.from(messageAttribute.Value, "base64").toString("ascii");
+            headers = JSON.parse(decodedValue);
+          }
 
-      let headers;
-      if (messageAttribute.Type === "String") {
-        headers = JSON.parse(messageAttribute.Value);
-      } else {
-        const decodedValue = Buffer.from(messageAttribute.Value, "base64").toString("ascii");
-        headers = JSON.parse(decodedValue);
+          const traceContext = this.tracerWrapper.extract(headers);
+          if (traceContext) {
+            logDebug("Extracted trace context from SNS-SQS event", { traceContext, event });
+            return traceContext;
+          } else {
+            logDebug("Failed to extract trace context from SNS-SQS event", { event });
+          }
+        }
+      }
+      // Then try to extract trace context from attributes.AWSTraceHeader. (Upstream Java apps can
+      // pass down Datadog trace context in the attributes.AWSTraceHeader in SQS case)
+      if (event?.Records?.[0]?.attributes?.AWSTraceHeader !== undefined) {
+        const traceContext = XrayService.extraceDDContextFromAWSTraceHeader(event.Records[0].attributes.AWSTraceHeader);
+        if (traceContext) {
+          logDebug("Extracted trace context from SNS-SQS event attributes.AWSTraceHeader", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("No Datadog trace context found from SNS-SQS event attributes.AWSTraceHeader", { event });
+        }
```

Comment:
Is there a way to make this code more readable?

I wonder if we could create functions inside this class for specific extraction scenarios, making the `extract` look more simple. WDYT?

Reply:
I actually think it's more readable to add the logging at this level to keep it consistent as above.
The `XrayService.extraceDDContextFromAWSTraceHeader` will just try to extract dd trace context if it exists or return null. 
"Returning if found, moving on if not" is more specific at this level here.
Let me know if it makes sense?